engine/sorted_collection/skiplist.cpp

```diff
@@ -1061,22 +1074,58 @@ void Skiplist::Destroy() {
 }
 
 void Skiplist::DestroyAll() {
-  GlobalLogger.Debug("Start Destroy skiplist with old version lists %s\n",
-                     Name().c_str());
+  GlobalLogger.Debug(
+      "Start Destroy skiplist with old version lists: %s, collection ID: %ld, "
+      "size: %ld\n",
+      Name().c_str(), ID(), Size());
   destroyAllRecords();
   destroyNodes();
-  GlobalLogger.Debug("Finish Destroy skiplist with old version lists %s\n",
-                     Name().c_str());
+  GlobalLogger.Debug(
+      "Finish Destroy skiplist with old version lists: %s, collection ID: "
+      "%ld\n",
+      Name().c_str(), ID());
 }
 
 void Skiplist::destroyNodes() {
   if (header_) {
-    SkiplistNode* to_delete = header_;
-    while (to_delete) {
-      SkiplistNode* next = to_delete->Next(1).RawPointer();
-      SkiplistNode::DeleteNode(to_delete);
-      to_delete = next;
+    // To avoid memory leak (don't free created skiplist node), we should
+    // iterate the skiplist to find all deleted skiplist nodes.
+    // Notice: sometimes, a thread has just marked deleted skiplist node A, but
+    // another thread found this deleted skiplist node A and updated its linkage
+    // of lower height whiling seeking nodes. It is easy to ignore that the
+    // linkage of higher height of this deleted skiplist node A has not been
+    // changed.
+    for (int i = header_->Height(); i >= 1; --i) {
+      auto prev = header_;
+      auto to_delete = header_->Next(i);
+      while (!to_delete.Null()) {
+        auto next = to_delete->Next(i);
+        auto cur_ptr = to_delete.RawPointer();
+        cur_ptr->valid_links--;
+        // Normally, the valid link numbers of skiplist node is 0 in 1th height.
```

Comment:
As you do valid_links-- for every nodes, we don't need to unlink anymore.
if(--cur_ptr->valid_links == 0){
deleteNode(cur_ptr);
}

Reply:
yes, only check `valid_links`  is zero.